### DISCUSSION TITLE
Request for Additional Attributes and Proposal for Unified Compatibility Package

### DISCUSSION THREAD

**Discussion by loop8ack** [manuelroemer/Nullable#31] [Ideas]
I find this package and `IsExternalInit` very handy and use them often. Thank you for that!
However, I still miss a few attributes. Are there plans to create packages for these?

- `System.Runtime.CompilerServices.CompilerFeatureRequiredAttribute` (`required`-Feature)
- `System.Runtime.CompilerServices.RequiredMemberAttribute` (`required`-Feature)
- `System.Runtime.CompilerServices.ModuleInitializerAttribute`
- `System.Runtime.CompilerServices.CallerArgumentExpressionAttribute`

I've frequently had to create these attributes manually and would find such packages extremely beneficial.

As a second thought:
I have always used both packages together, never separately. Wouldn't it be better to develop a joint package that can be expanded with additional content year after year? This package could then add the appropriate classes and members depending on the .NET version used. Of course, this should still be limited to simple things, such as attributes.

**Comment by manuelroemer** [comment]
> I find this package and IsExternalInit very handy and use them often. Thank you for that!

Thank you, that is nice to hear! ❤️ 

> However, I still miss a few attributes. Are there plans to create packages for these?
> [...]
> Wouldn't it be better to develop a joint package that can be expanded with additional content year after year? 

In principle, I am not be opposed to this, if there is a need for such packages. When I initially created Nullable a few years ago, it was one of the first packages providing such language polyfills. Nowadays, however, there exist projects like [PolySharp](https://github.com/Sergio0694/PolySharp) which kind of sound exactly like what you are asking for - one package to rule them all 😉. I don't consider Nullable / IsExternalInit obsolete or deprecated as they still work perfectly fine, but for large-ish projects that require the full scope of new language features, I'd typically recommend using PolySharp instead, if feasible.

My current plan for Nullable / IsExternalInit has therefore been to keep maintaining them (i.e., to fix issues and, in the case of Nullable, add new attributes should they ever be released). Regarding new attributes, however, I'd only consider creating new packages if existing solutions like PolySharp turned out to not fill the gap in certain situations or project configurations (which, so far, hasn't happened).

**Comment by loop8ack** [comment]
I didn't know PolySharp yet, that actually sounds like exactly what I'm looking for.

Thanks for the tip, I'll use that then :)
